src/transformStateWithClones.js

```diff
@@ -6,8 +6,79 @@
  *
  * @return {Object[]}
  */
+
 function transformStateWithClones(state, actions) {
-  // write code here
+  const stateHistory = [];
+
+  let currentState = { ...state };
+
+  actions.forEach((action) => {
+    currentState = { ...currentState };
+
+    if (action.type === 'addProperties') {
+      Object.assign(currentState, action.extraData);
+      stateHistory.push({ ...currentState });
+    }
+
+    if (action.type === 'removeProperties' && action.keysToRemove) {
+      action.keysToRemove.forEach((key) => {
+        delete currentState[key];
+      });
+      stateHistory.push({ ...currentState });
+    }
+
+    if (action.type === 'clear') {
+      currentState = {};
+      stateHistory.push({ ...currentState });
+    }
```

Comment:
Instead of multiple if statements, use a switch statement to handle the different `action.type` values. This will make the code cleaner and more maintainable. Remember to include a default case for error handling as well.